src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/SemanticTokens.cs

```diff
@@ -49,49 +49,105 @@ internal partial class RazorCustomMessageTarget
             return new ProvideSemanticTokensResponse(tokens: null, hostDocumentSyncVersion: csharpDoc.HostDocumentSyncVersion ?? -1);
         }
 
-        using var _ = ListPool<Task<ReinvocationResponse<VSSemanticTokensResponse>?>>.GetPooledObject(out var requestTasks);
-
         semanticTokensParams.TextDocument.Uri = csharpDoc.Uri;
         var textBuffer = csharpDoc.Snapshot.TextBuffer;
+        var response = semanticTokensParams.Ranges.Length == 1 ?
```

Comment:
Why not just check the feature flag? Having only one range is valid in both scenarios, and if the old scenario is still run sometimes even with the feature flag on, then the telemetry is not going to be telling the real story.

Plus making this method recursive is odd.

Reply:
I realized today that our unit tests are mocking the LanguageServerClient code, which means that any logic here in SemanticTokens.cs is not guarded by the tests.

- We use the real C# server to get the csharpTokens and setting up the mocked client to return those C# tokens.
- To improve the code, I am going to make `SemanticTokens.cs` as lean as possible, and have it be just a wrapper around the call to roslyn endpoint.
- This means that I will have two endpoints on the LanguageServerClient inside `SemanticTokens.cs`
- One is a wrapper around the existing endpoint, the other is a wrapper around the new endpoint.
- This way the recursive portion will also be removed

Reply:
Sounds like a great plan!